### ISSUE TITLE
[API Proposal]: Provide an API to get the native module handle of the native process entry-point module

### ISSUE THREAD

**Issue by jkoritzinsky** [dotnet/runtime#56331]
### Background and motivation

As mentioned in #7267, there is a strong desire for enabling using the "__Internal" name in a DllImport to point to methods in the hosting executable module of the process in both CoreCLR and Mono. Mono has had this support for many years, but CoreCLR doesn't have built-in support, and we don't have a desire to provide built-in support for this "hidden" name.

However, we understand the need for a way to resolve this name to the correct API. We propose adding an API that gets the native module handle of the entry-point module.

### API Proposal

```C#
namespace System.Runtime.InteropServices
{
     public static class NativeLibrary
     {
          public static IntPtr GetEntryPointModuleHandle();
     }
}
```     

Alternatively, we considered allowing `null` as a parameter to `NativeLibrary.Load(string)` to match how the implementation on non-Windows is `dlopen(NULL)`. The interop team decided that design was an abstraction leak and undesirable since the Windows implementation doesn't follow the same model. Additionally, the other `NativeLibrary.Load` overloads have additional behaviors and aren't thin wrappers, which makes the `null` option less desirable.

### API Usage

```C#
NativeLibrary.SetDllImportResolver(assembly, (name, _, paths) =>
{
     if (name == "__Internal")
     {
          return NativeLibrary.GetEntryPointModuleHandle();
     }
     return IntPtr.Zero;
});
```     

### Risks

None

**Comment by jkoritzinsky** [comment]
cc: @AaronRobinsonMSFT @elinor-fung

**Comment by GrabYourPitchforks** [comment]
Implementation question: how does this method affect the handle's reference count?

**Comment by jkoritzinsky** [comment]
This method will cache the handle after it's been called once, so it will only increase the refcount of the handle by 1. On Windows, we'll use `GetModuleHandle` which doesn't increase the refcount at all.

Since the module is the process's entry-point module, we don't need to worry about releasing the handle since it is valid until process tear-down (which will release the handle anyway).

**Comment by am11** [comment]
> Alternatively, we considered allowing `null` as a parameter to `NativeLibrary.Load(string)` to match how the implementation on non-Windows is `dlopen(NULL)`. 

Another approach is to return `(IntPtr)-1` which works on Windows and Unix alike.

https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/coreclr/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.CoreCLR.cs#L250

**Comment by lambdageek** [comment]
> Alternatively, we considered allowing null as a parameter to `NativeLibrary.Load(string)` to match how the implementation on non-Windows is `dlopen(NULL)`.

nit: this doesn't work on every unix.  In particular on Android `dlopen (NULL)` doesn't work, but `dlopen (RTLD_DEFAULT)` does.

**Comment by bartonjs** [comment]
Looks good as proposed.

```C#
namespace System.Runtime.InteropServices
{
     public static class NativeLibrary
     {
          public static IntPtr GetEntryPointModuleHandle();
     }
}
```

**Comment by jkotas** [comment]
> GetEntryPointModuleHandle

Module is very Windows-specific name. There is nothing called `Module` in the existing NativeLibrary APIs. Would `GetMainProgramHandle` be a better cross-platform name?

I assume that the non-Windows implementation is going to return `dlopen(NULL)`. Is that right? Note that `dlopen(NULL)` is a pseudo handle that does the lookup in all libraries loaded at startup, it does not correspond to a single file or module.

**Comment by jkoritzinsky** [comment]
I didn't want to use `GetMainProgramHandle` since that sounds like a process handle, not a library handle that's usable with `dlsym`

**Comment by jkoritzinsky** [comment]
And since the entrypoint of the program isn't a shared library, the name `GetMainProgramLibraryHandle` felt wrong as well.

**Comment by AaronRobinsonMSFT** [comment]
> Would GetMainProgramHandle be a better cross-platform name?

I don't have a strong preference here. "GetMainProgramHandle" or "GetEntryProgramHandle" are fine with me.

> Note that dlopen(NULL) is a pseudo handle that does the lookup in all libraries loaded at startup, it does not correspond to a single file or module.

I was not aware of that behavior. The latest doc on [`dlopen`](https://pubs.opengroup.org/onlinepubs/9699919799/functions/dlopen.html) makes this clear. Do we want this behavior or an actual handle to the main program?

>I didn't want to use GetMainProgramHandle since that sounds like a process handle, not a library handle that's usable with dlsym

That is fair but isn't a `NativeLibrary` scenario what we are attempting to support. After all, I believe the point of this API is to enable `"__Internal"` semantics. This would imply the primary use case of the returned value is for calls to `NativeLibrary.GetExport()`, no?

**Comment by jkoritzinsky** [comment]
> > I didn't want to use GetMainProgramHandle since that sounds like a process handle, not a library handle that's usable with dlsym
> 
> That is fair but isn't a `NativeLibrary` scenario what we are attempting to support. After all, I believe the point of this API is to enable `"__Internal"` semantics. This would imply the primary use case of the returned value is for calls to `NativeLibrary.GetExport()`, no?

Yes, we are trying to support a `NativeLibrary` scenario. My argument is that the new naming proposals imply a different scenario (process handle) than the scenario we are actually supporting and may cause confusion.

**Comment by AaronRobinsonMSFT** [comment]
> imply a different scenario (process handle) than the scenario we are actually supporting and may cause confusion.

Understood. I don't see this as a large concern given we can document the usage clearly and this is a niche scenario so minor confusion is okay and easily remedied. Not shutting down the conversation about renaming simply saying I don't think that will be that hard of a problem to address, but I do think that @jkotas's argument is fair - "Module" is a Windows-centric term.

Perhaps @bartonjs or @lambdageek have thoughts on naming alternatives?

**Comment by lambdageek** [comment]
> Perhaps @bartonjs or @lambdageek have thoughts on naming alternatives?

(I'm not great at naming) The man pages from Apple and Linux both call the value passed to `dlsym` a "handle" and they mention that there are some "pseudo-handles" and in particular `RTLD_DEFAULT` is the "default search order pseudo handle". So...

```csharp
namespace System.Runtime.InteropServices
{
     public static class NativeLibrary
     {
          public static IntPtr GetDefaultSearchOrderPsuedoHandle();
     }
}
```

**Comment by bartonjs** [comment]
> Module is very Windows-specific name.

During the API review there was theorization of a companion API like `IntPtr GetModuleHandle(System.Reflection.Module)`.  Because of Assembly.Modules/Assembly.ManifestModule/etc I think that "Module" is also a .NET term.  (It has much more affinity to .NET than to Windows for me, FWIW)

**Comment by jkotas** [comment]
Yes, `Module` is also .NET metadata term. I do not think we would want to introduce API like `IntPtr GetModuleHandle(System.Reflection.Module)` where the returned IntPtr would be the OS library handle. The metadata module is backed by Windows OS module only on CoreCLR, only in some cases, and it changed over time. This API would be leaking internal implementation details for no good reason.
